File: rxbinding-appcompat-v7/src/main/java/com/jakewharton/rxbinding2/support/v7/widget/SearchViewQueryTextChangeEventsObservable.java
package com.jakewharton.rxbinding2.support.v7.widget;

import android.support.v7.widget.SearchView;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class SearchViewQueryTextChangeEventsObservable
    extends InitialValueObservable<SearchViewQueryTextEvent> {
  private final SearchView view;

  SearchViewQueryTextChangeEventsObservable(SearchView view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super SearchViewQueryTextEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setOnQueryTextListener(listener);
  }

  @Override protected SearchViewQueryTextEvent getInitialValue() {
    return SearchViewQueryTextEvent.create(view, view.getQuery(), false);
  }

  final class Listener extends MainThreadDisposable implements SearchView.OnQueryTextListener {
    private final SearchView searchView;
    private final Observer<? super SearchViewQueryTextEvent> observer;

    Listener(SearchView searchView, Observer<? super SearchViewQueryTextEvent> observer) {
      this.searchView = searchView;
      this.observer = observer;
    }

    @Override public boolean onQueryTextChange(String s) {
      if (!isDisposed()) {
        observer.onNext(SearchViewQueryTextEvent.create(view, s, false));
        return true;
      }
      return false;
    }

    @Override public boolean onQueryTextSubmit(String query) {
      if (!isDisposed()) {
        observer.onNext(SearchViewQueryTextEvent.create(view, view.getQuery(), true));
        return true;
      }
      return false;
    }

    @Override protected void onDispose() {
      searchView.setOnQueryTextListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/widget/SearchViewQueryTextChangeEventsObservable.kt
@file:JvmName("RxSearchView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.widget

import android.widget.SearchView
import android.widget.SearchView.OnQueryTextListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding3.InitialValueObservable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of [query text events][SearchViewQueryTextEvent] on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun SearchView.queryTextChangeEvents(): InitialValueObservable<SearchViewQueryTextEvent> {
  return SearchViewQueryTextChangeEventsObservable(this)
}

private class SearchViewQueryTextChangeEventsObservable(
  private val view: SearchView
) : InitialValueObservable<SearchViewQueryTextEvent>() {

  override fun subscribeListener(observer: Observer<in SearchViewQueryTextEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.setOnQueryTextListener(listener)
    observer.onSubscribe(listener)
  }

  override val initialValue get() = SearchViewQueryTextEvent(view, view.query, false)

  private class Listener(
    private val view: SearchView,
    private val observer: Observer<in SearchViewQueryTextEvent>
  ) : MainThreadDisposable(), OnQueryTextListener {

    override fun onQueryTextChange(s: String): Boolean {
      if (!isDisposed) {
        observer.onNext(SearchViewQueryTextEvent(view, s, false))
        return true
      }
      return false
    }

    override fun onQueryTextSubmit(query: String): Boolean {
      if (!isDisposed) {
        observer.onNext(SearchViewQueryTextEvent(view, query, true))
        return true
      }
      return false
    }

    override fun onDispose() {
      view.setOnQueryTextListener(null)
    }
  }
}
